Funding signed event

jbesraa opened the pull request:
#3023 

Still a draft as yet to be fully implemented. channels currently cant progress from `FundingSigned`.

benthecarman commented on lightning/src/events/mod.rs:
```diff
@@ -497,6 +497,27 @@ impl_writeable_tlv_based_enum!(PaymentFailureReason,
+		/// [`ChannelManager::create_channel`]: crate::ln::channelmanager::ChannelManager::create_channel
+		/// [`ChannelManager::accept_inbound_channel`]: crate::ln::channelmanager::ChannelManager::accept_inbound_channel
+		/// [`UserConfig::manually_accept_inbound_channels`]: crate::util::config::UserConfig::manually_accept_inbound_channels
+		user_channel_id: u128,
```
You should give the funding tx to broadcast

jbesraa commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2864,6 +2865,23 @@ macro_rules! send_channel_ready {
 }
+macro_rules! emit_funding_signed_event {
+	($locked_events: expr, $channel: expr) => {
+		let funding_tx = $channel.context.get_funding_transaction(); // hmm.. this is always none?
```
the funding_transaction is `None` while it should be defined at this stage as we are after `FundingCreated`.. what could be the reason for this?

jbesraa commented on lightning/src/ln/channel.rs:
```diff
@@ -646,6 +646,7 @@ mod state_flags {
 	pub const LOCAL_SHUTDOWN_SENT: u32 = 1 << 11;
 	pub const SHUTDOWN_COMPLETE: u32 = 1 << 12;
 	pub const WAITING_FOR_BATCH: u32 = 1 << 13;
+	pub const FUNDING_SIGNED: u32 = 1 << 14;
```
`FUNDING_SIGNED` is added as part of `AwaitingChannelReadyFlags` and not a whole new thing

jbesraa commented:
Done few things:
1. Added #3056  here 
2. Removed `FUNDING_SIGNED` channel state
3. Removed config `manually_broadcast_..` argument
4. Other things like using `clone` here https://github.com/lightningdevkit/rust-lightning/pull/3024/files#diff-8d28e96b6f27edb9a9739b2dce334f8906ec906155d421a7b5b02aa8e6c96057R5276 

Yet to do:
1. Complete the docs
2. Fix CI
3. Improve/fix naming around the new property in `ChannelContext`

jbesraa commented:
@TheBlueMatt @benthecarman this is ready for review

jbesraa commented on lightning/src/ln/channel.rs:
```diff
@@ -1547,6 +1550,14 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
+	/// The funding transaction can be accessed through the [`Event::FundingSigned`] event.
+	///
+	/// [`Event::FundingSigned`]: crate::events::Event::FundingSigned
+	manually_broadcast_outbound_channels: Option<()>,
```
The `Option<()>` is mainly because other variables `is_batch_funding` and `local_initiated_shutdown` are defined with `Option` so I was trying to follow that.

jbesraa commented on lightning/src/ln/channel.rs:
```diff
@@ -2365,13 +2389,25 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider  {
 	// Checks whether we should emit a `ChannelPending` event.
 	pub(crate) fn should_emit_channel_pending_event(&mut self) -> bool {
-		self.is_funding_broadcast() && !self.channel_pending_event_emitted
+		self.is_funding_broadcast() && !self.channel_pending_event_emitted && self.manually_broadcast_outbound_channels.is_none()
```
I think we have two options here:
1) Let the user tell us they broadcasted the transaction and as a result we emit `ChannelPending`
2) Track the chain and check when the transaction is in the mempool and emit `ChannelPending`

First option is not always doable from the user perspective, for example in payjoin scenario you dont really know when the tx will be broadcasted.  wdyt?

jbesraa replied:
That would be the easiest if we are comfortable emitting both events simultaneously

jbesraa commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2865,6 +2865,23 @@ macro_rules! send_channel_ready {
 }
+macro_rules! emit_funding_signed_event {
+	($locked_events: expr, $channel: expr) => {
+		let should_emit = $channel.context.should_emit_funding_signed_event();
```
Following your suggestion above with emitting both events simultaneously, could we have the same `should_emit_..`(roughly, excluding the respective variable `is_emitted..`) function for `FundingSigned` and `ChannelPending` events?

jbesraa commented:
Thanks for the reviews, addressed all the comments.

jbesraa commented:
Thanks @tnull 
Fixed-up the following:
```shell
diff --git a/lightning/src/events/mod.rs b/lightning/src/events/mod.rs
index 38077c8b..a2403c3e 100644
--- a/lightning/src/events/mod.rs
+++ b/lightning/src/events/mod.rs
@@ -535,7 +535,7 @@ pub enum Event {
        /// Used to indicate that the counterparty node has provided the signature(s) required to
        /// recover our funds in case they go offline.
        ///
-       /// It is safe(and your responsibility) to broadcast the funding transaction upon receiving this
+       /// It is safe (and your responsibility) to broadcast the funding transaction upon receiving this
        /// event.
        ///
        /// This event is only emitted if you called
@@ -1489,7 +1489,7 @@ impl Writeable for Event {
                                        (2, user_channel_id, required),
                                        (4, funding_tx, required),
                                        (6, counterparty_node_id, required),
-                                       (8, former_temporary_channel_id, required),
+                                       (8, former_temporary_channel_id, option),
                                });
                        },
                        // Note that, going forward, all new events must only write data inside of
@@ -1939,7 +1939,7 @@ impl MaybeReadable for Event {
                                                (2, user_channel_id, required),
                                                (4, funding_tx, required),
                                                (6, counterparty_node_id, required),
-                                               (8, former_temporary_channel_id, required)
+                                               (8, former_temporary_channel_id, option)
                                        });
                                        Ok(Some(Event::FundingTxBroadcastSafe {
                                                channel_id: channel_id.0.unwrap(),
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index 9d39ca72..ac316d45 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -9352,7 +9352,7 @@ impl<'a, 'b, 'c, ES: Deref, SP: Deref> ReadableArgs<(&'a ES, &'b SP, u32, &'c Ch
                                // Later in the ChannelManager deserialization phase we scan for channels and assign scid aliases if its missing
                                outbound_scid_alias: outbound_scid_alias.unwrap_or(0),

-                               funding_tx_broadcast_safe_event_emitted: funding_tx_broadcast_safe_event_emitted.unwrap_or(true),
+                               funding_tx_broadcast_safe_event_emitted: funding_tx_broadcast_safe_event_emitted.unwrap_or(false),
                                channel_pending_event_emitted: channel_pending_event_emitted.unwrap_or(true),
                                channel_ready_event_emitted: channel_ready_event_emitted.unwrap_or(true),

@@ -9365,7 +9365,7 @@ impl<'a, 'b, 'c, ES: Deref, SP: Deref> ReadableArgs<(&'a ES, &'b SP, u32, &'c Ch
                                local_initiated_shutdown,

                                blocked_monitor_updates: blocked_monitor_updates.unwrap(),
-                               is_manual_broadcast: is_manual_broadcast.unwrap_or(true),
+                               is_manual_broadcast: is_manual_broadcast.unwrap_or(false),
                        },
                        #[cfg(any(dual_funding, splicing))]
                        dual_funding_channel_context: None,
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index f6816708..d1b773b9 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -4339,27 +4339,23 @@ where
        }


-       /// Unsafe: This method does not check if the funding transaction is signed ie. if the witness data
+       /// Unsafe: This method does not check if the funding transaction is signed, i.e., if the witness data
        /// is empty or not. It is the caller's responsibility to ensure that the funding transaction
        /// is final.
        ///
-       /// If you wish to use a safer method, use [`ChannelManager::funding_transaction_generated`]
+       /// If you wish to use a safer method, use [`ChannelManager::funding_transaction_generated`].
        ///
        /// Call this upon creation of a funding transaction for the given channel.
        ///
        /// Note that if this method is called successfully and the counterparty sent us
-       /// `funding_signed`, the funding transaction wont be broadcasted and you are expected to
-       /// broadcast it manually when receiving the `FundingTxBroadcastSafe` event.
+       /// `funding_signed` msg, the funding transaction won't be broadcasted and you are expected to
+       /// broadcast it manually when receiving the [`FundingTxBroadcastSafe`] event.
        ///
-       /// In contrary to the original `funding_transaction_generated` method, the `ChannelPending`
-       /// event is never emitted. Once the funding transaction is confirmed by both parties, the
-       /// next event will be `ChannelReady`.
-       ///
-       /// Returns an [`APIError::APIMisuseError`] if the funding_transaction spent non-SegWit outputs
+       /// Returns an [`APIError::APIMisuseError`] if the `funding_transaction` spent non-SegWit outputs
        /// or if no output was found which matches the parameters in [`Event::FundingGenerationReady`].
        ///
        /// Returns [`APIError::APIMisuseError`] if the funding transaction is not final for propagation
-       /// across the p2p network.
+       /// across the peer-to-peer network.
        ///
        /// Returns [`APIError::ChannelUnavailable`] if a funding transaction has already been provided
        /// for the channel or if the channel has been closed as indicated by [`Event::ChannelClosed`].
@@ -4368,7 +4364,7 @@ where
        /// channel (note that this should be trivially prevented by using unique funding transaction
        /// keys per-channel).
        ///
-       /// Note that this includes RBF or similar transaction replacement strategies - lightning does
+       /// Note that this includes RBF or similar transaction replacement strategies - Lightning does
        /// not currently support replacing a funding transaction on an existing channel. Instead,
        /// create a new channel with a conflicting funding transaction.
        ///
@@ -4378,6 +4374,7 @@ where
        /// for more details.
        ///
        /// [`Event::FundingGenerationReady`]: crate::events::Event::FundingGenerationReady
+       /// [`Event::FundingTxBroadcastSafe`]: crate::events::Event::FundingTxBroadcastSafe
        /// [`Event::ChannelClosed`]: crate::events::Event::ChannelClosed
        /// [`ChannelManager::funding_transaction_generated`]: crate::ln::channelmanager::ChannelManager::funding_transaction_generated
        pub fn unsafe_manual_funding_transaction_generated(&self, temporary_channel_id: &ChannelId, counterparty_node_id: &PublicKey, funding_transaction: Transaction) -> Result<(), APIError> {
```

jbesraa commented:
A usage example of this PR can be seen in this commit: https://github.com/lightningdevkit/ldk-node/pull/301/commits/d137df1d81fdd6b52d8d68c9876a3f5e3434e406

specifically,  `channel_manager::unsafe_manual_funiding_..` is called here 
https://github.com/lightningdevkit/ldk-node/commit/d137df1d81fdd6b52d8d68c9876a3f5e3434e406#diff-03a28135b3f471a98f1c40eb8a4dcc7c4759f4bf9aabc83c12c71f14d68e00c1R228

and `FundingTxBroadcastSafe` is handled here 
https://github.com/lightningdevkit/ldk-node/commit/d137df1d81fdd6b52d8d68c9876a3f5e3434e406#diff-14da7c350de5742eaa248de110fefeae9b4b639c0358e26024d6040ccf174a41R937

jbesraa commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4321,6 +4338,53 @@ where
 	}
 
+
+	/// Unsafe: This method does not check if the funding transaction is signed, i.e., if the witness data
+	/// is empty or not. It is the caller's responsibility to ensure that the funding transaction
```
hmm the only difference between this function and the original `funding_transaction_generated` is the fact that the transaction can be NOT-signed in this 'unsafe' variant. a comment about segwit can be seen a couple of lines below

jbesraa replied:
lol, yea sorry. removed that.

jbesraa replied:
Is this clear?
```shell
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -4339,9 +4339,8 @@ where
        }


-       /// Unsafe: This method does not check if the funding transaction is signed, i.e., if the
-       /// witness data is empty or not. It is the caller's responsibility to ensure that the funding
-       /// transaction is final.
+       /// Unsafe: This method does not check if the funding transaction spend non-SegWit outputs. It is the
+       /// caller's responsibility to ensure the funding transaction validatiy.
        ///
        /// If you wish to use a safer method, use [`ChannelManager::funding_transaction_generated`].
```

jbesraa commented on lightning/src/events/mod.rs:
```diff
@@ -532,6 +532,40 @@ pub enum Event {
+		counterparty_node_id: PublicKey,
+		/// The `temporary_channel_id` this channel used to be known by during channel establishment.
+		///
+		/// Will be `None` for channels created prior to LDK version 0.0.115.
```
this would lead to this `unwrap` https://github.com/lightningdevkit/rust-lightning/pull/3024/commits/ed0812603df113c854280edbd3ea8560c05a4838#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cR2559

jbesraa commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4321,6 +4338,53 @@ where
+	/// [`Event::FundingTxBroadcastSafe`]: crate::events::Event::FundingTxBroadcastSafe
+	/// [`Event::ChannelClosed`]: crate::events::Event::ChannelClosed
+	/// [`ChannelManager::funding_transaction_generated`]: crate::ln::channelmanager::ChannelManager::funding_transaction_generated
+	pub fn unsafe_manual_funding_transaction_generated(&self, temporary_channel_id: &ChannelId, counterparty_node_id: &PublicKey, funding_transaction: Transaction) -> Result<(), APIError> {
```
yea I think its saved and then `.take()` somewhere 
there is also `lock_time` checking inside `batch_funding_transaction_generated_intern`.. should we go with it as is?

jbesraa commented:
@TheBlueMatt 
Cherry-picked your commit here. It does indeed improve readability and provide more flexibility but does make most of whats returned from `FundingGenerationReady` event redundant. I am cool with either.

jbesraa commented:
Thanks @tnull. Squashed commits with the following diff

```shell
On branch funding-signed-event
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   lightning/src/events/mod.rs
        modified:   lightning/src/ln/channel.rs
        modified:   lightning/src/ln/channelmanager.rs

no changes added to commit (use "git add" and/or "git commit -a")
[I] ecode@pop-os: rust-lightning [funding-signed-event]× » git diff .                                                               [14:55:56]
diff --git a/lightning/src/events/mod.rs b/lightning/src/events/mod.rs
index f135bbd20..a81394a6b 100644
--- a/lightning/src/events/mod.rs
+++ b/lightning/src/events/mod.rs
@@ -596,7 +596,7 @@ pub enum Event {
                channel_id: ChannelId,
                /// The `user_channel_id` value passed in to [`ChannelManager::create_channel`] for outbound
                /// channels, or to [`ChannelManager::accept_inbound_channel`] for inbound channels if
-               /// [`UserConfig::manually_accept_inbound_channels`] config flag is set to true. Otherwise
+               /// [`UserConfig::manually_accept_inbound_channels`] config flag is set to `true`. Otherwise
                /// `user_channel_id` will be randomized for an inbound channel.
                ///
                /// [`ChannelManager::create_channel`]: crate::ln::channelmanager::ChannelManager::create_channel
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index 08396b440..2cb36deed 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -1331,13 +1331,11 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
        counterparty_forwarding_info: Option<CounterpartyForwardingInfo>,

        pub(crate) channel_transaction_parameters: ChannelTransactionParameters,
-       /// The transaction which funds this channel. Note that for manually-funded channels(i.e,
+       /// The transaction which funds this channel. Note that for manually-funded channels (i.e.,
        /// is_manual_broadcast is true) this will be a dummy empty transaction.
        funding_transaction: Option<Transaction>,
        /// This flag indicates that it is the user's responsibility to validated and broadcast the
        /// funding transaction.
-       ///
-       /// [`Event::FundingTxBroadcastSafe`]: crate::events::Event::FundingTxBroadcastSafe
        is_manual_broadcast: bool,
        is_batch_funding: Option<()>,

@@ -8803,8 +8801,8 @@ impl<SP: Deref> Writeable for Channel<SP> where SP::Target: SignerProvider {
                        (45, cur_holder_commitment_point, option),
[I] ecode@pop-os: rust-lightning [funding-signed-event]× » git status                                                               [14:57:22]
On branch funding-signed-event
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   lightning/src/events/mod.rs
        modified:   lightning/src/ln/channel.rs
        modified:   lightning/src/ln/channelmanager.rs

no changes added to commit (use "git add" and/or "git commit -a")
[I] ecode@pop-os: rust-lightning [funding-signed-event]× » git diff .                                                               [14:57:23]
diff --git a/lightning/src/events/mod.rs b/lightning/src/events/mod.rs
index f135bbd20..a81394a6b 100644
--- a/lightning/src/events/mod.rs
+++ b/lightning/src/events/mod.rs
@@ -596,7 +596,7 @@ pub enum Event {
                channel_id: ChannelId,
                /// The `user_channel_id` value passed in to [`ChannelManager::create_channel`] for outbound
                /// channels, or to [`ChannelManager::accept_inbound_channel`] for inbound channels if
-               /// [`UserConfig::manually_accept_inbound_channels`] config flag is set to true. Otherwise
+               /// [`UserConfig::manually_accept_inbound_channels`] config flag is set to `true`. Otherwise
                /// `user_channel_id` will be randomized for an inbound channel.
                ///
                /// [`ChannelManager::create_channel`]: crate::ln::channelmanager::ChannelManager::create_channel
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index 08396b440..2cb36deed 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -1331,13 +1331,11 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
        counterparty_forwarding_info: Option<CounterpartyForwardingInfo>,
 
        pub(crate) channel_transaction_parameters: ChannelTransactionParameters,
-       /// The transaction which funds this channel. Note that for manually-funded channels(i.e,
+       /// The transaction which funds this channel. Note that for manually-funded channels (i.e.,
        /// is_manual_broadcast is true) this will be a dummy empty transaction.
        funding_transaction: Option<Transaction>,
        /// This flag indicates that it is the user's responsibility to validated and broadcast the
        /// funding transaction.
-       ///
-       /// [`Event::FundingTxBroadcastSafe`]: crate::events::Event::FundingTxBroadcastSafe
        is_manual_broadcast: bool,
        is_batch_funding: Option<()>,
 
@@ -8803,8 +8801,8 @@ impl<SP: Deref> Writeable for Channel<SP> where SP::Target: SignerProvider {
                        (45, cur_holder_commitment_point, option),
diff --git a/lightning/src/events/mod.rs b/lightning/src/events/mod.rs
index f135bbd20..a81394a6b 100644
--- a/lightning/src/events/mod.rs
+++ b/lightning/src/events/mod.rs
@@ -596,7 +596,7 @@ pub enum Event {
                channel_id: ChannelId,
                /// The `user_channel_id` value passed in to [`ChannelManager::create_channel`] for outbound
                /// channels, or to [`ChannelManager::accept_inbound_channel`] for inbound channels if
-               /// [`UserConfig::manually_accept_inbound_channels`] config flag is set to true. Otherwise
+               /// [`UserConfig::manually_accept_inbound_channels`] config flag is set to `true`. Otherwise
                /// `user_channel_id` will be randomized for an inbound channel.
                ///
                /// [`ChannelManager::create_channel`]: crate::ln::channelmanager::ChannelManager::create_channel
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index 08396b440..2cb36deed 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -1331,13 +1331,11 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
        counterparty_forwarding_info: Option<CounterpartyForwardingInfo>,
 
        pub(crate) channel_transaction_parameters: ChannelTransactionParameters,
-       /// The transaction which funds this channel. Note that for manually-funded channels(i.e,
+       /// The transaction which funds this channel. Note that for manually-funded channels (i.e.,
        /// is_manual_broadcast is true) this will be a dummy empty transaction.
        funding_transaction: Option<Transaction>,
        /// This flag indicates that it is the user's responsibility to validated and broadcast the
        /// funding transaction.
-       ///
-       /// [`Event::FundingTxBroadcastSafe`]: crate::events::Event::FundingTxBroadcastSafe
        is_manual_broadcast: bool,
        is_batch_funding: Option<()>,
 
@@ -8803,8 +8801,8 @@ impl<SP: Deref> Writeable for Channel<SP> where SP::Target: SignerProvider {
                        (45, cur_holder_commitment_point, option),
                        (47, next_holder_commitment_point, option),
                        (49, self.context.local_initiated_shutdown, option), // Added in 0.0.122
-                       (51, is_manual_broadcast, option),
-                       (53, funding_tx_broadcast_safe_event_emitted, option)
+                       (51, is_manual_broadcast, option), // Added in 0.0.124
+                       (53, funding_tx_broadcast_safe_event_emitted, option) // Added in 0.0.124
                });
                                                                                                                                              
                Ok(())
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 033e1b730..b8f241381 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -2620,7 +2620,8 @@ macro_rules! emit_funding_tx_broadcast_safe_event {
                                user_channel_id: $channel.context.get_user_id(),
                                funding_txo: $funding_txo,
                                counterparty_node_id: $channel.context.get_counterparty_node_id(),
-                               former_temporary_channel_id: $channel.context.temporary_channel_id().expect("Unreachable: FundingTxBroadcastSafe event feature added to channel establishment process in LDK v0.124.0 where this should never be None."),
+                               former_temporary_channel_id: $channel.context.temporary_channel_id()
+                                       .expect("Unreachable: FundingTxBroadcastSafe event feature added to channel establishment process in LDK v0.0.124 where this should never be None."),
                        }, None));
                        $channel.context.set_funding_tx_broadcast_safe_event_emitted();
                }
@@ -4434,7 +4435,7 @@ where
        }
                                                                                                                                              
                                                                                                                                              
-       /// Unsafe: This method does not check the validatiy of the provided output. It is the
+       /// **Unsafe**: This method does not check the validity of the provided output. It is the
        /// caller's responsibility to ensure that.
        ///
        /// For a safer method, please refer to [`ChannelManager::funding_transaction_generated`].
(END)
```

jbesraa commented:
@TheBlueMatt please see last commit for changes.

I am not sure now we still need `funding_transaction_generated_intern` and `funding_transaction_generated` ? wdyt?

jbesraa commented:
@TheBlueMatt Addressed all comments
